Updates to README.md

degville opened the pull request:
Updated README.md to include new build instructions and updated branch details. 

cc @stulluk

degville committed:
updated README.md to include new build instructions and updated branch info.

degville committed:
small edit

degville committed:
typo

degville committed:
changed arch to reflect example.

stulluk commented:
Thank you @degville , I believe this is a good progress.

Would it be better to improve "building" part as follows:

# Building

You can build this gadget snap in two ways: **native build** and **cross build**

## Cross building on your host

### Prequisities
- Any ubuntu host ( 20.04 or newer is recommended )
- snapcraft
- multipass

### Building with a multipass virtual machine
 Just execute **"snapcraft"** in the root directory of this git repository. This will create&run a multipass vm, and build this snap inside this vm, putting output file in the current directory.

### Building in a lxd container

Just execute **"snapcraft --use-lxd"**

## Native Building on your Raspberry-Pi

### Prequisities
- Any supported Rpi variant with Ucore20+ installed
- Network connection via SSH
- Internet access of Rpi device

### Building on Rpi
Run following commands on your RPi ( access to your pi via shell / ssh)

```
$ sudo snap install lxd
$ sudo lxd init --auto
$ sudo lxc launch ubuntu:20.04 focal
$ sudo lxc shell focal
# snap install snapcraft --classic
# git clone https://github.com/snapcore/pi-gadget/
# cd pi-gadget
# snapcraft --destructive-mode
```

This will build pi-gadget.snap natively on your RPi

----------------

For your consideration.

degville committed:
updates after GH feedback.

degville committed:
typo. Added sentence.

degville committed:
typo.

xnox commented:
@stulluk surely on Rpi native, if one has lxd & snapcraft installed one can simply do `snapcraft --provider lxd` no?

Without a need to launch a focal lxd container mode themselves and use destructive mode.

Or does `snapcraft --provider lxd` not work for native builds for some reason?

degville commented:
@stulluk I've updated the README.md to hopefully address your comments: I trimmed the Cross building section slightly but left the arch details and example to hopefully help. I also added your suggested section on native building on LXD - it would be good to know what this approach might be best for as it's likely going to be slower than the native build version.

Re: not using `--provider lxd`, I assumed it was because we needed git, which isn't available on UC20. Happy to update this with comments and feedback.

xnox commented:
> @stulluk I've updated the README.md to hopefully address your comments: I trimmed the Cross building section slightly but left the arch details and example to hopefully help. I also added your suggested section on native building on LXD - it would be good to know what this approach might be best for as it's likely going to be slower than the native build version.
> 
> Re: not using `--provider lxd`, I assumed it was because we needed git, which isn't available on UC20. Happy to update this with comments and feedback.

Sadness. I wonder if snapcraft snap should stage git inside it, and use it if git is not available on the host. @sergiusens does that make sense ?

stulluk commented:
> @stulluk surely on Rpi native, if one has lxd & snapcraft installed one can simply do `snapcraft --provider lxd` no?
> 
> Without a need to launch a focal lxd container mode themselves and use destructive mode.
> 
> Or does `snapcraft --provider lxd` not work for native builds for some reason?

On my pi:

```
stulluk@ubuntu:~$ snap install snapcraft
error: snap "snapcraft" requires classic confinement which is only available on classic systems
stulluk@ubuntu:~$ 
```

How to install snapcraft natively on the Rpi with Ucore ? Sorry for my ignorance.

xnox commented:
> > @stulluk surely on Rpi native, if one has lxd & snapcraft installed one can simply do `snapcraft --provider lxd` no?
> > Without a need to launch a focal lxd container mode themselves and use destructive mode.
> > Or does `snapcraft --provider lxd` not work for native builds for some reason?
> 
> On my pi:
> 
> ```
> stulluk@ubuntu:~$ snap install snapcraft
> error: snap "snapcraft" requires classic confinement which is only available on classic systems
> stulluk@ubuntu:~$ 
> ```
> 
> How to install snapcraft natively on the Rpi with Ucore ? Sorry for my ignorance.

you are right, it's not even possible to install snapcraft hence yeah, one needs to spawn a classical environment somehow, hence lxd and focal classic inside lxd.

I somewhat wish we could use snapcraft as a strictly confined snap =( especially with the lxd provider.

stulluk commented:
> @stulluk I've updated the README.md to hopefully address your comments: I trimmed the Cross building section slightly but left the arch details and example to hopefully help. I also added your suggested section on native building on LXD - it would be good to know what this approach might be best for as it's **likely going to be slower than the native build** version.
> 
> Re: not using `--provider lxd`, I assumed it was because we needed git, which isn't available on UC20. Happy to update this with comments and feedback.

Considering 
1) performance enhancements on SBC's in the last 10 years
2) availability of the SBC's not only for daily usage but also as development environments 

IMHO, we should not restrict readers to large customers, I think it would be nicer if tinkerers/hobbyists/makers.. and a poor child living in nowhere in middle east should be able to build their applications on low cost devices. Everybody should be able to benefit from our products. 

IMHO, The documentation should not make assumptions on the knowledge or experience level of the readers. It should be written in a language which is understandable, readable and and cover all possible use-cases ( not hundred pages long, but adequate )

The code itself should also be consistent with the documentation, which is sadly not.

There are multiple bugs in the snapcraft.yaml in this repository, which avoid this snap to be built via cross compilation and @ogra will create a PR whenever he finds time, or I will.

degville commented:
> IMHO, we should not restrict readers to large customers, I think it would be nicer if tinkerers/hobbyists/makers.. and a poor child living in nowhere in middle east should be able to build their applications on low cost devices. Everybody should be able to benefit from our products.
> 
> IMHO, The documentation should not make assumptions on the knowledge or experience level of the readers. It should be written in a language which is understandable, readable and and cover all possible use-cases ( not hundred pages long, but adequate )

I totally agree with all of this. If someone is using a Pi as their main machine, though, I think it would be safe to assume they're running either Pi Ubuntu Server/Desktop or Raspberry Pi OS rather than Ubuntu Core (at least currently). In which case, the normal Snapcraft instructions should apply (though in my experience snapcraft is problematic on RPi OS).

sergiusens commented:
> I somewhat wish we could use snapcraft as a strictly confined snap =( especially with the lxd provider.

we could eventually work on a snapd interface that allows it to do what you would do most things required on the host.

degville committed:
added command to pull snap from LXC.

anonymouse64 commented on README.md:
```diff
@@ -39,25 +39,108 @@ in `psplash/config` to point to this file and rebuild the gadget.
+building](https://ubuntu.com/core/docs/board-enablement#heading--image-building) 
+for instructions on how to build a bootable image that includes the gadget snap.
+
+### Native building
```
I don't mean to nitpick, but I think really we could make this more clear by separating how you build not by native vs cross, but rather by _snapcraft-managed_ or _manual-managed_ build environments. 

Here's where the various builds fall into these two categories for example:
- build the gadget snap on a classic ubuntu amd64 desktop -> snapcraft managed 
- build the gadget snap on a classic ubuntu arm64 pi desktop -> snapcraft managed 
- build the gadget snap on an ubuntu core pi machine -> manual managed via lxd (or multipass maybe, I dunno if KVM support on pi is working yet, it's not the past few times I checked)
- build the gadget snap on some fast specific build machine -> manual managed via lxd or whatever

Then the tutorial essentially has your `Cross building` section become the snapcraft-managed build environment where you install the classic snapcraft snap + either lxd or multipass and just do `snapcraft` and get a working snap, or you describe how folks can setup a manually managed instance via LXD like one would do on UC20 (or wherever they want to run LXD), pointing out the most important bits that 1) the managed instance has the same distro series as declared in the gadget's snapcraft.yaml `base` setting, which will change depending on if you are building the 20 gadget or the 18 gadget for example and 2) that snapcraft is called with `--destructive-mode` flag.

anonymouse64 replied:
Hi, did anyone consider this sort of refactoring to this doc?

anonymouse64 reviewed (commented):
I think we can organize this a bit better, the content looks good, but splitting it up between cross building and native building isn't as useful IMHO as splitting it up between building with _snapcraft-managed_ build instances and _manually-managed_ build instances

anonymouse64 commented:
> we could eventually work on a snapd interface that allows it to do what you would do most things required on the host.

@sergiusens it would be cool if snapcraft had strict and classic tracks, where the strict track does less things, like doesn't do destructive-mode for example and only uses LXD, etc. but that's probably a bit of work and kinda makes the track situation a bit complicated since y'all will have other version number based tracks too

sergiusens commented:
> @sergiusens it would be cool if snapcraft had strict and classic tracks, where the strict track does less things, like doesn't do destructive-mode for example and only uses LXD, etc. but that's probably a bit of work and kinda makes the track situation a bit complicated since y'all will have other version number based tracks too

Instead of that I would maybe instead focus on having a nice interface for snapcraft to _do things_. I once upon a time proposed dual confinement (a snap that defaults to strict but allows for --classic) `confinement: [strict, classic]`, a special interface we can leverage can also be something to look into.

xnox commented on README.md:
```diff
@@ -1,36 +1,36 @@
+This repository contains the source for an [Ubuntu
+Core](https://ubuntu.com/core) gadget snap that runs universally on all the
+Raspberry Pi boards currently supported by Ubuntu Core (Raspberry Pi 2B, 3B,
+3A+, 3B+, 4B, Compute Module 3, and Compute Module 3+).
```
Compute module 4 too, no?

sil2100 reviewed (approved):
Ok, I think this is a good starting point. I think we can get this merged and then address issues mentioned by both Dimitri and Ian.

degville committed:
updates to README.md after GH feedback.

anonymouse64 commented on README.md:
```diff
@@ -53,10 +54,35 @@ the two Raspberry Pi architectures:
+Another managed option is to run `snapcraft remote-build`. This command
+offloads the snap build process to the [Launchpad build
+farm](https://launchpad.net/builders), pushing the potentially
+multi-architecture snap back to your machine when the build completes. See
```
```suggestion
foreign architecture snap back to your machine when the build completes. See
```

degville replied:
Thanks!

anonymouse64 commented on README.md:
```diff
@@ -53,10 +54,35 @@ the two Raspberry Pi architectures:
+Examples of both a _managed_ build and a _manually managed_ build are outlined
+below.
+
+### Example managed build
```
just a comment re-reading the below section again, while managed is the most convenient method, it is not the most performant, the most performant is manually managed since there is not all the extra snapcraft overhead of managing the VM/container, since you are managing the stuff yourself when building

degville replied:
Thanks for taking a look! I've tweaked the text slightly to remove this.

anonymouse64 reviewed (approved):
This looks really great, thanks for re-writing the section

degville committed:
After GH feedback, removed misleading performant statement, plus other small edit.

waveform80 commented:
Taking a look at merging this.

Personally, I don't like the change from "native/cross build" to the "managed" headers; the former are meaningful without a grounding in snapcraft. For example, "native" or "cross build" would help me skip directly to the section I want; I probably know in advance whether I want to build on a Pi (which might be my only option) or a PC, and those headings help me skip directly to the instructions I need. By contrast the "managed" headers don't mean much to me and I have to read a bunch of prose to determine what section I need to follow.

Having said that: I'm no expert on snapcraft, everyone else seems happy with the headings, and I'd much rather have all this extra documentation than not -- it's *much* better than what went before. Thanks very much for the contribution!

I'll get this merged in a mo and cherry-pick it across all the {20,22}-arm{hf,64} branches. I'm not going to merge it into the classic or desktop branches for obvious reasons (different build process), but do let me know if it should go into the 18- branches too (I don't know whether these instructions would work with those branches).
